compiler/passes/src/flattening/flattener.rs

```diff
@@ -90,35 +137,146 @@ impl<'a> Flattener<'a> {
         }
     }
 
-    /// Clears the state associated with `ReturnStatements`, returning the ones that were previously stored.
-    pub(crate) fn clear_early_returns(&mut self) -> Vec<(Option<Expression>, ReturnStatement)> {
-        core::mem::take(&mut self.returns)
+    /// Construct an early return guard.
+    ///
+    /// That is, an Identifier assigned to a boolean that is true iff some early return was taken.
+    pub(crate) fn construct_early_return_guard(&mut self) -> Option<(Identifier, Vec<Statement>)> {
+        if self.returns.is_empty() {
+            return None;
+        }
+
+        if self.returns.iter().any(|g| matches!(g.0, ReturnGuard::None)) {
+            // There was a return with no conditions, so we should simple return True.
+            let place = Identifier {
+                name: self.assigner.unique_symbol("true", "$"),
+                span: Default::default(),
+                id: self.node_builder.next_id(),
+            };
+            let statement = self.simple_assign_statement(
+                place,
+                Expression::Literal(Literal::Boolean(true, Default::default(), self.node_builder.next_id())),
+            );
+            return Some((place, vec![statement]));
+        }
+
+        // All guards up to a certain point in the stack should be constructed.
+        // Find the first unconstructed one.
+        let start_i = (0..self.returns.len())
+            .rev()
+            .take_while(|&i| matches!(self.returns[i].0, ReturnGuard::Unconstructed(_)))
+            .last()
+            .unwrap_or(self.returns.len());
+
+        let mut statements = Vec::with_capacity(self.returns.len() - start_i);
+
+        for i in start_i..self.returns.len() {
+            let ReturnGuard::Unconstructed(identifier) = self.returns[i].0 else {
```

Comment:
Nit. As a general practice, if we use `unreachable!` we should add a comment justifying why.